Initial version

bobsilverberg opened the pull request:
This is the initial version of FxAPOM. 

As can be seen from the readme, it includes a method for creating a FxA, as well as a helper which can be used to sign in via FxA.  I am looking for feedback on this before making the initial commit to Mozilla/master. I will also be releasing this as a PyPI package when it has passed review.

Note that this code is already being run in production by marketplace-tests and marketplace-tests-gaia, both pointing at my repo.

@davehunt r?
@AndreiH r?

AndreiH commented on .travis.yml:
```diff
@@ -0,0 +1,18 @@
+  - 2.6
+  - 2.7
+
+script: py.test --baseurl=https://123done-stable.dev.lcip.org/ --driver=firefox --webqatimeout=90 --destructive -m "not credentials" tests
```
I think the url is the wrong one. Shouldn't it be <code>https://stable.dev.lcip.org/</code>?

bobsilverberg replied:
No, this is the url for the testing site, not for the FxA API.  If you run the tests locally you'll see what I mean.

davehunt commented on .travis.yml:
```diff
@@ -0,0 +1,18 @@
+before_script:
+  - sh -e /etc/init.d/xvfb start
```
Add "export DISPLAY=:99.0" before this line and remove the 'env' section from below. I'd also suggest making this 'before_install' and adding and install section for flake8, and a 'before_script' for running flake8.

bobsilverberg replied:
Done

davehunt commented on .travis.yml:
```diff
@@ -0,0 +1,18 @@
+  email:
+    - webqa-ci@mozilla.org
+  irc:
+    - "irc.mozilla.org#mozwebqa"
```
Nit: These could be on the same line, rather than a list with a single item.

bobsilverberg replied:
Done

davehunt commented on conftest.py:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
```
Nit: This shebang is in a lot of our files. As I understand it's only needed if you're going to run the file directly from a terminal. I'd like to remove them whenever they're unnecessary.

bobsilverberg replied:
Done. Removed from all files.

davehunt commented on conftest.py:
```diff
@@ -0,0 +1,16 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from selenium.webdriver.support.ui import WebDriverWait
```
Nit: Add a blank line after the license.

bobsilverberg replied:
Done.

davehunt commented on credentials.yaml:
```diff
@@ -0,0 +1,3 @@
+default:
+    email:
+    password:
```
Can we avoid using a credentials file by generating Firefox accounts?

bobsilverberg replied:
Removed this file. We don't need it.

davehunt commented on fxapom/fxapom.py:
```diff
@@ -0,0 +1,59 @@
+class FxATestAccount:
+    """A base test class that can be extended by other tests to include utility methods."""
+
+    password = ''.join([random.choice(string.letters) for i in xrange(8)])
```
Why `xrange` and not `range`?

bobsilverberg replied:
I'm not sure. This was in the code when I started making my mods. From what I can understand, `xrange` is potentially faster, but that shouldn't matter in this case. I also understand that `xrange` is not Python 3 compatible, so I will change it to `range`.

davehunt commented on fxapom/fxapom.py:
```diff
@@ -0,0 +1,59 @@
+        self.account = TestEmailAccount(email=email_pattern)
+        client = Client(self.fxa_url)
+        print 'fxapom created an account for email: %s at %s on %s' % (
+            self.account.email, self.fxa_url, datetime.now())
```
Should this statement be after the account is created?

bobsilverberg replied:
Good point. Moved.

davehunt commented on fxapom/pages/sign_in.py:
```diff
@@ -0,0 +1,77 @@
+            self.popup = True
+            for handle in self.selenium.window_handles:
+                self.selenium.switch_to.window(handle)
+                if self.is_element_present(*self._fox_logo_locator):
```
We're using the presence of an element with ID 'fox-logo' to determine if we're in the correct window. This feels a little fragile. Is there something else we could use. When I was looking into signing in via Marionette I used `'Sign in' in title`.

bobsilverberg replied:
I tried that initially and the login screen used by Marketplace, which is the only suite currently using this tool, behaves strangely in that the title switches from "Firefox Accounts" to "Sign in ....", so there were timing issues. I tried a number of things and this was the most reliable one I came up with.

davehunt replied:
Fair enough - perhaps raise an issue against Firefox Account to get something like `<body class="sign-in">` or similar.

davehunt commented on fxapom/pages/sign_in.py:
```diff
@@ -0,0 +1,77 @@
+        password.send_keys(value)
+
+    def click_next(self, expect='password'):
+        self.selenium.find_element(*self._next_button_locator).click()
```
We're not using the expect argument at all. Is it needed?

bobsilverberg replied:
Good catch. I think that was a leftover.

davehunt commented on readme.md:
```diff
@@ -1 +1,54 @@
-readme
\ No newline at end of file
+**F**irefo**x** **A**ccounts **P**age **O**bject **M**odel
```
This styling is not being applied, probably because it's a heading.

bobsilverberg replied:
Removed.

davehunt commented on readme.md:
```diff
@@ -1 +1,54 @@
\ No newline at end of file
+**F**irefo**x** **A**ccounts **P**age **O**bject **M**odel
+===============================================
+Webdriver compatible page object model and utilities for Firefox Accounts
```
Can we make WebDriver and Firefox Account links?

bobsilverberg replied:
Done.

davehunt commented on readme.md:
````diff
@@ -1 +1,54 @@
+for `FxATestAccount`, or, if you know you want to create a production Account, you can omit that argument.
+
+Example:
+```
````
Use ```python for syntax highlighting.

bobsilverberg replied:
Done

davehunt commented on readme.md:
````diff
@@ -1 +1,54 @@
+Example:
+```
+from fxapom.fxapom import WebDriverFxA
+fxa = WebDriverFxA(mozwebqa)
````
We should probably note that using this will only work with pytest-mozwebqa. It might be nice to support passing in the selenium instance and optional timeout for anyone not using our plugin.

bobsilverberg replied:
Added a note about this being applicable to the plugin. I will leave adding optional args for the constructor to a future enhancement request, if one ever comes in.

davehunt commented on readme.md:
```diff
@@ -1 +1,54 @@
+
+Running The Tests
+-----------------
+* Install the requirements using `pip install -r requirements.txt`
```
pip install works for me

bobsilverberg replied:
Yeah, that error above, @AndreiH, wouldn't seem to be related to running `setup`. I'm not sure what the issue is you are seeing.

davehunt commented on readme.md:
```diff
@@ -1 +1,54 @@
+Running The Tests
+-----------------
+* Install the requirements using `pip install -r requirements.txt`
+* Run the tests via `py.test tests`
```
We should also specify a `--driver` or browser/platform details.

bobsilverberg replied:
Done

davehunt commented on setup.py:
```diff
@@ -0,0 +1,26 @@
+      keywords='mozilla',
+      author='Mozilla Web QA',
+      author_email='mozwebqa@mozilla.org',
+      url='https://github.com/bobsilverberg/fxapom',
```
URL should be to the mozilla fork

bobsilverberg replied:
Done

davehunt commented on tests/test_create_account.py:
```diff
@@ -0,0 +1,31 @@
+from fxapom.fxapom import FxATestAccount
+
+
+@pytest.mark.nondestructive
```
These are creating accounts, are they non-destructive?

bobsilverberg replied:
The accounts are being created via the test account server, so I kind of see them as nondestructive, but I see your point. Removing the mark will still allow the tests to run against the dev environment.

davehunt commented on tests/test_login.py:
```diff
@@ -0,0 +1,33 @@
+    _fxa_logged_in_indicator_locator = (By.ID, 'loggedin')
+
+    @pytest.mark.credentials
+    def test_existing_user_can_sign_in(self, mozwebqa):
```
You may have to remind me, what's the difference between these two tests? One is using an account we create, the other is using an existing account, but what's the benefit of using an existing account?

bobsilverberg replied:
Another good point. I initially saw this as two different test cases, but what we are testing, in terms of this package, is identical. To have these two tests is really to be testing FxA itself, which is unnecessary.

davehunt commented:
Thanks @bobsilverberg I like this a lot! I can't see any blockers for getting this merged. All of my comments can be addressed in followups if they're considered worthwhile.

bobsilverberg commented:
Thanks for the review @davehunt. I have pushed a new commit which addresses all of your comments, except for those that I responded to feeling a change was not warranted. Would you mind doing another quick r?

davehunt commented:
B.e.a.utiful! r+

bobsilverberg commented:
Thanks @davehunt!
